components/merkledb/src/views/owning_iter.rs

```diff
@@ -0,0 +1,337 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{marker::PhantomPinned, mem, pin::Pin};
+
+use crate::{
+    db::Iterator,
+    views::{RawAccess, View},
+};
+
+#[derive(Debug)]
+struct Inner<T: RawAccess> {
+    // Safety: Not accessed during iteration.
+    view: View<T>,
+    iter: Option<Box<dyn Iterator>>,
+    _pin: PhantomPinned,
+}
+
+impl<T: RawAccess> Inner<T> {
```

Comment:
To be honest I like neither of two existing solutions. On the one hand, loading all the group keys into RAM is dangerous, and on the other the solution which I don't find very elegant.

Taking into account that in most cases you will not iterate through group keys often, maybe it's possible to create an intermediate approach:
- Define some fixed amount of entries that we can safely store in RAM;
- Using `View::iter` load the first bunch of entries and allow iterating through them;
- Once all the loaded entries are consumed, load the next portion using `View::iter_from`;
- Finally, stop iterating, when all the entries with the given prefix are consumed.

Will this work or is it even more inappropriate?

Reply:
I've thought about this approach briefly. It is incorrect in the general case (the keys can change between `iter_from` calls), but I *think* it could work for readonly access, which is the use case for migrations. I'll try this approach, most probably in a separate branch (since this one has owning iterators as the defining feature).